app/src/main/java/de/westnordost/streetcomplete/screens/about/LogsViewModelImpl.kt

```diff
@@ -0,0 +1,62 @@
+package de.westnordost.streetcomplete.screens.about
+
+import androidx.lifecycle.viewModelScope
+import de.westnordost.streetcomplete.data.logs.LogMessage
+import de.westnordost.streetcomplete.data.logs.LogsController
+import de.westnordost.streetcomplete.data.logs.LogsFilters
+import de.westnordost.streetcomplete.util.ktx.systemTimeNow
+import de.westnordost.streetcomplete.util.ktx.toEpochMilli
+import de.westnordost.streetcomplete.util.ktx.toLocalDate
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.getAndUpdate
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.LocalTime
+
+class LogsViewModelImpl(
+    private val logsController: LogsController
+) : LogsViewModel() {
+
+    private val logsControllerListener = object : LogsController.Listener {
+        override fun onAdded(message: LogMessage) {
+            if (filters.value.matches(message)) {
+                // TODO this is hugely inefficient (log list is copied every time a single entry is added!!!)
+                logs.update { it + message }
```

Comment:
To solve this, we can replace `logs` with a `MutableSharedFlow` with a replay of 1 and have a private `ArrayList` as the backing. Whenever the `ArrayList` is appended to, you can `emit` the `ArrayList` into `MutableSharedFlow` (which does not care about equation of the emitted values)

Reply:
Ah, nice idea. I didn't foray into `SharedFlow` yet, but it seems quite straightforward.

So, like this?

https://github.com/streetcomplete/StreetComplete/pull/5482/commits/314f05f75df208a543a3566a5c81f6a31731ae8d

Reply:
Hm, and this also does not solve the issue really. Yes, the array list is not copied in the view model anymore, but in the LogsFragment it is. The `LogsAdapter` needs to know whether just one item was added at the end or the whole list was replaced, because, well, that's how these `RecyclerView.Adapter`s work, with `notifyItemInserted` etc.

Maybe this isn't an issue in Compose, I don't know.

(Anyway, I will revert the linked commit for now)